WASM version of Animation System.

gonnavis opened the pull request:
Sibling PR: https://github.com/upstreet-labs/app/pull/116

Follow: https://github.com/webaverse/app-wasm/pull/63 , but this time try to keep exactly the same logics as js version, without refactoring. Except for in `_handleDefault()` function, using `blendList()` method instead of `_get7wayBlend()`.

Tested with updating animations 200 times per frame, wasm version get almost 40 fps, vs js version around 25 fps.
_( Please notice that chrome dev tool has big influence on performance, so need close it when testing performance. )_

200 loop:
js: ~ 25 fps            | wasm: ~ 40 fps
:-------------------------:|:-------------------------:
![image](https://user-images.githubusercontent.com/10785634/189923085-a43e06cf-d578-4ac3-9ec8-d356452e2966.png)   | ![image](https://user-images.githubusercontent.com/10785634/191592561-a2c21f74-b079-4dc1-a8e8-8bcaf923575d.png)

gonnavis commented:
- Cleaned codes and logics.

- I'm still using animation groups infos defined in js side, and transfer to wasm side.
  I think it's fine for the first version.
  Will try to move them into `animation-groups.json` or `animations-baker.mjs` later.

- For the `cubicBezier` function porting, I can get `y` by `t`, but has problem get `y` by `x`, haven't found existing cpp implementation yet.
  I saw js version codes used precalculated lookup table to get `y` by `x`, and used so many techniques to prove performance:
  https://github.com/gre/bezier-easing
  > Implementing efficient lookup is not easy because it implies projecting the X coordinate to a Bezier Curve. This micro library uses fast heuristics (involving dichotomic search, newton-raphson, sampling) to focus on performance and precision.

  Will keep trying to find cpp implementation, or implement them in wasm by self if really needed.
  Now I'm using `pow(x, 0.1)` instead temporarily, their function curve are very similar:
  https://cubic-bezier.com/#0,1,0,1
  ![image](https://user-images.githubusercontent.com/10785634/192064309-9badbc46-9ebb-44af-bd36-949cfab87e0a.png)
  https://www.mathsisfun.com/calculus/derivative-plotter.html
  ![image](https://user-images.githubusercontent.com/10785634/192064483-27f3ce02-c038-41a5-92f9-3d8b67e0a651.png)

avaer commented on AnimationSystem/AnimationSystem.h:
```diff
@@ -0,0 +1,137 @@
+    // values
+    std::string activateAnimationName, fallLoopFrom, sitAnimationName, danceAnimationName, emoteAnimationName, useAnimationName, useAnimationComboName, hurtAnimationName, unuseAnimationName, aimAnimationName, holdAnimationName;
+    std::vector<std::string> useAnimationEnvelopeNames;
+    float activateTime, landTime, fallLoopFactor, fallLoopTime, flyTime, doubleJumpTime, jumpTime, narutoRunTime, narutoRunTimeFactor, danceFactor, crouchMaxTime, emoteFactor, lastEmoteTime, idleWalkFactor, useTime, useAnimationEnvelopeLength, hurtTime, unuseTime, aimTime, aimMaxTime, walkRunFactor, crouchFactor, pickUpTime, forwardFactor, backwardFactor, leftFactor, rightFactor, mirrorLeftFactor ...
```
Probably best on separate lines for when we make changes.

gonnavis replied:
Updated https://github.com/webaverse/app-wasm/pull/75/commits/873179145010e397f0152674bfcc5b7dc9e9f138 .

avaer commented on AnimationSystem/AnimationSystem.h:
```diff
@@ -0,0 +1,137 @@
+    static float nowS;
+
+    Avatar *avatar;
+    float *animationValues[53]; // 53 bones interpolants result buffers.
```
Probably best to get this from a compile-time declaration. Eg. `sizeof()` some array which defines all of the possible bones.

Otherwise when we update it is very likely we will make a mistake and cause crashes.

gonnavis replied:
Updated https://github.com/webaverse/app-wasm/pull/75/commits/1963f6c8ac4a37ae8ab3107fd6f2c5cd4a556342 .

avaer reviewed (commented):
This looks quite good and readable.

The one suggestion I would make is not use string maps at runtime. The animation group names look like they are defined in this code already, so it could be array indexes.

If we animate a lot of avatars, looking up the animation strings every frame will probably make a difference.

gonnavis commented on AnimationSystem/AnimationSystem.h:
```diff
@@ -0,0 +1,137 @@
+  class AnimationNode;
+  class AnimationMixer;
+
+  enum LoopType {
```
Oh, not used, sorry I missed cleaning this up. Will delete.

gonnavis replied:
Updated https://github.com/webaverse/app-wasm/pull/75/commits/1c0894cb3883df0aa592390fb7e6e5f0a30d754f .

avaer commented:
For the cubic bezier, we can probably port https://github.com/webaverse/app/blob/master/easing.js to C++. It shouldn't be too hard.

gonnavis commented on AnimationSystem/AnimationSystem.h:
```diff
@@ -0,0 +1,137 @@
+    LoopRepeat = 2201,
+    LoopPingPong = 2202
+  };
+  enum BoneIndex {
```
Updated https://github.com/webaverse/app-wasm/pull/75/commits/29ca9a1b84438e6444e6f7bd93c0527f46895415 .

gonnavis commented:
> For the cubic bezier, we can probably port https://github.com/webaverse/app/blob/master/easing.js to C++. It shouldn't be too hard.

Ported https://github.com/webaverse/app-wasm/compare/540ef872dc4fb0aa19ba6eb03516228735071bf0...f29b5b9d5170d9d5d1d2b1ac8d76e921106f4669

gonnavis commented:
> The one suggestion I would make is not use string maps at runtime. The animation group names look like they are defined in this code already, so it could be array indexes.
> 
> If we animate a lot of avatars, looking up the animation strings every frame will probably make a difference.

Used `enum`s + `std::vector`s instead of `string`s + `std::map`s.
https://github.com/webaverse/app-wasm/compare/f29b5b9d5170d9d5d1d2b1ac8d76e921106f4669...8a301a809cf32605c96cf72b6d4c587143e1e4eb
https://github.com/upstreet-labs/app/compare/b3ec43eb2f4febe57052104e20e55f823a6c33f8...c53b207191ede0967858168a97679f2d7543e69b

Really a big performance improvement again!
With 200 loop, now get full 60 fps.
With 1000 loop, get ~ 28 fps, still better than js version 200 loop ~ 25 fps.

Current WASM version:
200 loop, full 60 fps            | 1000 loop, ~ 28 fps
:-------------------------:|:-------------------------:
![image](https://user-images.githubusercontent.com/10785634/192629363-40012571-524e-4bec-9e15-b02b66cd14dd.png)   | ![image](https://user-images.githubusercontent.com/10785634/192629557-3d54d8bd-a357-49be-95ee-34c9bfde8f3c.png)

avaer commented:
> Really a big performance improvement again!
With 200 loop, now get full 60 fps.

That's crazy, wasn't expecting that...

avaer commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1098 @@
+
+      // -------------------------------------------------------------------------
+
+      // note: Need be the same order with `constants.h/enum ActivateAnimationIndex`.
```
This not is good, but...

Is it possible to iterate the enum determinisitcally so we don't have to maintain 2 places in code.

avaer commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1098 @@
+
+      // -------------------------------------------------------------------------
+
+      // note: Need be the same order with `constants.h/enum ActivateAnimationIndex`.
```
This note is good, but...

Is it possible to iterate the enum determinisitcally so we don't have to maintain 2 places in code?

gonnavis replied:
Used `initializer list`s instead of `enum`s, to put animation groups' declarations in one place, all in `constants.h` now.

https://github.com/webaverse/app-wasm/compare/8a301a809cf32605c96cf72b6d4c587143e1e4eb...ab2b22d3e2c43874a4f6645648d9e36b1b316d1a

gonnavis commented:
If only playing narutoRun animation, can get ~ 28 fps under 3000 loop.
![image](https://user-images.githubusercontent.com/10785634/193066179-8f71e799-3fa8-499d-b8e8-2f0cf2a60a16.png)

avaer commented on AnimationSystem/constants.h:
```diff
@@ -0,0 +1,844 @@
+enum BoneIndex {
```
Can we use IOTA for this so there are no explicit numbers?

gonnavis replied:
OK, I'll change.

gonnavis replied:
Used existing bone indexes infos, instead of explicit numbers and iota.
https://github.com/webaverse/app-wasm/pull/75/commits/7c6b01bbd80c001a5f46711ecfde36ebaa2000ae
https://github.com/webaverse/app-wasm/pull/75/commits/8d30b38e277531c626023eebcf041c3dfa3fd5a6

avaer commented on AnimationSystem/AnimationSystem.h:
```diff
@@ -0,0 +1,133 @@
+    std::string name;
+  };
+  struct AnimationMapping { // spec
+    float *dst =  (float *)malloc((4) * sizeof(float));
```
What's this for? Not generally good to have this random malloc()...

gonnavis replied:
It's used for keep animation result data like what js version's `animationMapping.dst` do.
Because of `animationMappings` only need init once globally, and will return these `dst` values back to js, so I allocated them.
https://github.com/webaverse/app-wasm/pull/75/files#diff-71459e596906bb9318622660fe80e75db679de846febc40fc79f6f81e48a7ed3R1069
Let me try if there's nicer way to do it.

gonnavis replied:
For performance reason, I directly return `dst`s' pointers to js side.
https://github.com/webaverse/app-wasm/blob/808c4102b91357d9f31cdc793f2f5bf9236eb6eb/AnimationSystem/AnimationSystem.cc#L1234-L1237
So I think have to use `heap` memory instead of `stack` memory.
But I used `new float[size]` instead of `(float *)malloc((size) * sizeof(float))` for clearer codes.
https://github.com/webaverse/app-wasm/pull/75/commits/808c4102b91357d9f31cdc793f2f5bf9236eb6eb
Is it ok?

avaer replied:
No, I still have no idea what this is about. There should never be a reason to initialize struct members to the heap like this.

If you need a global, use a global. Otherwise use an inline member like an array.

avaer replied:
From your other code I'm assuming you intended to use an inline array. Maybe you are confused about how to read it in JS?

gonnavis replied:
Used inline array and global singleton instead of initialize struct members to the heap.
https://github.com/webaverse/app-wasm/compare/b667920cc2bbda7da0c901cf2236cdf2c3c99e8d...b56cd596f08bc8147647384576cab9b44352f961

avaer replied:
I don't get it. Why does it need `malloc`?

avaer commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1074 @@
+
+      // -------------------------------------------------------------------------
+
+      defaultSitAnimationIndex = sitAnimationIndexes.Chair;
```
These defaults should not exist in C++. They must be computed on the JS side and always passed to C++.

If you don't explicitly pass the index you want, that's an error.

gonnavis replied:
https://github.com/webaverse/app-wasm/pull/75/files#diff-bfaaa55a4ec2614935143fb5dabf5097733ba995063b5fec95020a26bf65c7b9R713-R727
Because of now defining these animations infos directly in C++, not in JS.
So I think we can't and don't need compute on JS side then pass to C++.
And then will never pass wrong index.

avaer commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1074 @@
+
+      // -------------------------------------------------------------------------
+
+      activateSpeedFactors.push_back(1.2); // Grab_forward
```
Since this is highly dependent on the animations, this code should not exist in the C++.

If the animations are the wrong speed then the animation should be updated.

gonnavis replied:
Updated animation `.fbx` files directly, instead of using `activateSpeedFactors`.
https://github.com/upstreet-labs/app/pull/116/commits/13b7f9c828e7c43a35037b6193733f9bc47c194b
https://github.com/upstreet-labs/app/pull/116/commits/b8a20a6d5fe47ac1e6f0f84fa9a07c5245f89152
https://github.com/webaverse/app-wasm/pull/75/commits/f99e1f672b9390aa5940634178286e17fa79f52b

avaer commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1074 @@
+
+      // -------------------------------------------------------------------------
+
+      jsonStr += "[";
```
Instead of writing your own JSON, I think there are better JSON libraries to that we can import that are less likely to error in corner cases, link non-utf8 data.

gonnavis replied:
I've successfully used https://github.com/nlohmann/json to parse JSON strings from JS side.
But I can't find `stringify` feature in it.
And very surprisingly that, because of type and reflection reason, it's very hard to find a JSON stringify C++ lib.
I'll keep investigating.

gonnavis replied:
Trying to use this lib
https://cppsecrets.com/users/13187108971001001049711497106484964103109971051084699111109/C00-PocoJSONObjectstringify.php
https://github.com/pocoproject
https://github.com/pocoproject/poco

gonnavis replied:
After tried many cpp json libs and failed to integrate ( mainly because of Undefined Symbol error ).
I investigated "nlohmann/json" again ( it's a single header cpp lib, very easy to integrate ).
And found it really has `stringify()` feature, but named `bump()` instead.
https://github.com/webaverse/app-wasm/compare/19d0d52831bc4ca52d7e768dc125bf477bbbb24e...29630e753c58e9aceb7a6fb0e09532f414cada20

avaer reviewed (commented):
The main issue remaining is that we should not have any animation-dependent constants in the C++, especially constants to rescale specific animations.

The C++ part should be simple and not have any special cases or awareness of the animations used. If there are defaults, they should be defaulted in the JS and always passed here, so that the control exists on the JS side.

And if we are rescaling animations here, then perhaps we should just rescale the animation and get rid of the scale constants here?

gonnavis commented on AnimationSystem/AnimationSystem.cc:
```diff
@@ -0,0 +1,1074 @@
+        jsonStr += "\"" + declaration.groupName + "\"";
+        jsonStr += ",";
+        jsonStr += "\"index\":";
+        jsonStr += std::to_string(declaration.index);
```
I use `to_string` to convert cpp `int` type to cpp `string` type so it can be added into `jsonStr`.
Not to define `string` type in JSON.
So the result is correctly `"index":0`, not `"index":"0"`.

gonnavis commented:
- Added most `ActionInterpolant`s directly on WASM side.
- Added WASM `addAction()`/`removeAction` apis.
- Calculate a lot of `xxxState`s, `xxxTime`s directly on WASM side, instead of get from JS.
https://github.com/upstreet-labs/app/compare/17c6762..c26aa80

gonnavis commented:
- Calculate all `actionInterpolants` on wasm side.
- Get some `actionInterpolants` back from wasm to js when needed.
- Deleted all js side `actionInterpolants`. ( For localPlayer/UninterpolatedPlayer )

https://github.com/webaverse/app-wasm/compare/c312bed..a2ac8a6
https://github.com/upstreet-labs/app/compare/17c6762..85da9c4
